### PR TITLE
Add pattern_end_offset to signature_state

### PR DESCRIPTION
**Pull request by ronny8360988**
Currently, when a signature is matched, a ```signature_match``` event is raised with the signature state, with the message and with the data as parameters, where ```data``` contains the whole payload of the last input chunk, and not just the matched payload.

This PR aims to add a ```pattern_end_offset``` field to ```signature_state```, such that ```signature_match``` events will have the context about the last offset within ```data``` in which the pattern is matched.

This field is optional, since not all signatures are matched against a pattern (i.e. pure rules).

### CODE REVIEW

**Review by timwoj — CHANGES_REQUESTED** [review]
This looks fine to me. The pre-commit CI build failing so that needs to be fixed. Can you squash it down to a single commit?

File: src/RuleMatcher.cc [hunk 1]
```diff
@@ -954,12 +958,14 @@ bool RuleMatcher::EvalRuleConditions(Rule* r, RuleEndpointState* state, const u_
...
 
-void RuleMatcher::ExecRuleActions(Rule* r, RuleEndpointState* state, const u_char* data, int len, bool eos) {
+void RuleMatcher::ExecRuleActions(Rule* r, RuleEndpointState* state, const u_char* data, int len, bool eos,
+                                  std::optional<int> match_end_offset) {
     if ( state->opposite && is_member_of(state->opposite->matched_rules, r->Index()) )
         // We have already executed the actions.
         return;
 
     state->matched_rules.push_back(r->Index());
+    state->match_end_offset = match_end_offset;
```
[the remark this answers is not in the document]
Comment (ronny8360988) [reply src/RuleMatcher.cc:968]: Great suggestion! Applying

**Comment by awelzel** [comment]
I think this looks good now. Any chance you could provide a test, too? If not, I'm happy to create one during merge. I'd also squash all commits down.

One question/note:

> This PR aims to add a pattern_end_offset field to signature_state, such that signature_match events will have the context about the last offset within data in which the pattern is matched.

This isn't quite what you'll get. `data` is going to be the most recent packet data that was fed into the signature framework, but offset could be relative to the stream start  with e.g. TCP, so `offset` might end up larger than `|data|` and can't be used for indexing into `data`. The offset might still be useful to see how deep into a stream the match was, but it doesn't always help to determine the offset in `data`.

**Comment by eladsolomon-ms** [comment]
> This isn't quite what you'll get. `data` is going to be the most recent packet data that was fed into the signature framework, but offset could be relative to the stream start with e.g. TCP, so `offset` might end up larger than `|data|` and can't be used for indexing into `data`. The offset might still be useful to see how deep into a stream the match was, but it doesn't always help to determine the offset in `data`.

WOW, you are totally correct!
What a bit oversight by us, thanks for pointing that out.

The intention here is to get the offset in `data`, for data extraction. How do you suggest we calculate that offset?

**Comment by awelzel** [comment]
> The intention here is to get the offset in data, for data extraction. How do you suggest we calculate that offset?

Think we could keep the offset on the RuleEndpointState and then compute relative ones for the matches. I've fiddled with it in the following branch, also added a test with a match in the second TCP packet of a stream:

https://github.com/zeek/zeek/commit/0669aa4c2d7e454d9917c8946cc5eb7e100ed4d4

Robin also brought up adding offset as another parameter to `signature_match()` - that's a bit awkward due to custom event support and the optional `msg` we have now, but should be possible. I'll give it another try.

**Comment by awelzel** [comment]
@ronny8360988 / @eladsolomon-ms - I'm going to close this PR, a compatible feature was added with #3996 .

The `signature_match()` event was extended to receive `end_of_match: count` which is relative to the beginning of `data`. This allows you to determine which part of `data` was still matched. Though, note that `data` only contains the current packet/payload, so anything matched in prior packets there's no visibility.

**Pull request closed without merging**